src/http/controllers/jet.rs

```diff
@@ -1,198 +1,147 @@
                 Ok(_) => { /* alright */ }
             }
         }
 
-        // create association
-        let mut jet_associations = self.jet_associations.lock().unwrap(); // no need to deal with lock poisoning
+        // Controller runs by Saphir via tokio 0.2 runtime, we need to use .compat()
+        // to run Mutex from tokio 0.3 via Saphir's tokio 0.2 runtime. This code should be upgraded
+        // when saphir perform transition to tokio 0.3
+        let mut jet_associations = self.jet_associations.lock().compat().await;
 
         jet_associations.insert(association_id, Association::new(association_id, JET_VERSION_V2));
-        start_remove_association_future(
-            self.executor_handle.clone(),
-            self.jet_associations.clone(),
-            association_id,
-        );
+        start_remove_association_future(self.jet_associations.clone(), association_id).await;
 
-        res.status(StatusCode::OK);
+        (StatusCode::OK.as_u16(), ())
     }
 
-    fn gather_association_candidates(&self, req: &SyncRequest, res: &mut SyncResponse) {
+    #[post("/association/<association_id>/candidates")]
+    async fn gather_association_candidates(&self, req: Request) -> (u16, Option<String>) {
         let association_id = match req
             .captures()
             .get("association_id")
             .and_then(|id| Uuid::parse_str(id).ok())
         {
             Some(id) => id,
-            None => {
-                res.status(StatusCode::BAD_REQUEST);
-                return;
-            }
+            None => return (StatusCode::BAD_REQUEST.as_u16(), None),
         };
 
         // check the session token is signed by our provider if unrestricted mode is not set
         if !self.config.unrestricted {
-            match validate_session_token(self.config.as_ref(), req) {
+            match validate_session_token(self.config.as_ref(), &req) {
                 Err(e) => {
                     slog_scope::error!("Couldn't validate session token: {}", e);
-                    res.status(StatusCode::UNAUTHORIZED);
-                    return;
+                    return (StatusCode::UNAUTHORIZED.as_u16(), None);
                 }
                 Ok(expected_id) if expected_id != association_id => {
                     slog_scope::error!(
                         "Invalid session token: expected {}, got {}",
                         expected_id,
                         association_id
                     );
-                    res.status(StatusCode::FORBIDDEN);
-                    return;
+                    return (StatusCode::FORBIDDEN.as_u16(), None);
                 }
                 Ok(_) => { /* alright */ }
             }
         }
 
         // create association
-        let mut jet_associations = self.jet_associations.lock().unwrap(); // no need to deal with lock poisoning
+        let mut jet_associations = self.jet_associations.lock().compat().await; // no need to deal with lock poisoning
```

Comment:
You can remove this comment since there is no lock poisoning check with tokio Mutex

Reply:
Yep, forgot to clean-up the comments, thanks!